apps/src/code-studio/progressRedux.js

```diff
@@ -208,116 +212,59 @@ export default function reducer(state = initialState, action) {
         viewType === ViewType.Student && state.studentDefaultsSummaryView
     };
   }
-
   if (action.type === SET_CURRENT_STAGE_ID) {
     // if we already have a currentStageId, that means we're on a puzzle page,
     // and we want currentStageId to remain the same (rather than reflecting
     // the last stage the user has made progress on).
     if (state.currentStageId) {
       return state;
     }
-
     return {
       ...state,
-      currentStageId: action.stageId
+      currentStageId: action.stageId,
+      currentLesson: getLesson(state, action.stageId)
     };
   }
-
   if (action.type === SET_SCRIPT_COMPLETED) {
     return {
       ...state,
       scriptCompleted: true
     };
   }
-
   if (action.type === SET_STAGE_EXTRAS_ENABLED) {
     return {
       ...state,
       stageExtrasEnabled: action.stageExtrasEnabled
     };
   }
-
   return state;
 }
 
 // Helpers
 
-/**
- * Return the level with the highest progress, or the first level if none have
- * been attempted
- * @param {number[]} levelIds
- * @param {Object.<number,number>} - Mapping from level id to progress result
- */
-function bestResultLevelId(levelIds, progressData) {
-  // The usual case
-  if (levelIds.length === 1) {
-    return levelIds[0];
-  }
-
-  // Return the level with the highest result
-  var attemptedIds = levelIds.filter(id => progressData[id]);
-  if (attemptedIds.length === 0) {
-    // None of them have been attempted, just return the first
-    return levelIds[0];
-  }
-  var bestId = attemptedIds[0];
-  var bestResult = progressData[bestId];
-  attemptedIds.forEach(function(id) {
-    var result = progressData[id];
-    if (result > bestResult) {
-      bestId = id;
-      bestResult = result;
+export const getLesson = (state, id) => {
+  let foundLesson;
+  for (const group of state.lessonGroups) {
+    for (const lesson of group.lessons) {
```

Comment:
NIT: If you'd like to save some keystrokes in the future, you can use `find` for this. i.e. `group.lessons.find(lesson => lesson.id === id)`

Reply:
ah yes. i tried to figure out how to do a nested find but there doesn't seem to be a way to do that, but you're right that i can still use it for the inner search!